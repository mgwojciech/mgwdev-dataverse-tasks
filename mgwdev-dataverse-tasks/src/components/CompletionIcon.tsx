import { makeStyles } from "@fluentui/react";
import * as React from "react";

export interface ICompletionIconProps {
    percentComplete: number;
}

const useStyles = makeStyles({
    root: {

    },
    wrapper: {
        width: "24px",
        height: "24px",
        borderRadius: "50%",
        border: "4px solid #eaeaea",
        position: "relative",
        transitionProperty: "all",
        transitionDuration: "0.3s",
        transitionTimingFunction: "ease-in-out",
        "&:hover": {
        }
    },
    circle: {
        position: "absolute",
        top: "-3px",
        left: "-3px",
        display: "block",
        width: "100%",
        height: "100%",
        borderRadius: "50%",
        border: "3px solid",
    }
})
export const CompletionIcon: React.FC<ICompletionIconProps> = (props) => {
    const classNames = useStyles();
    return <div className={classNames.root}>
        <div className={classNames.wrapper}>
            <div className={classNames.circle} style={{
                borderColor: `rgba(${255 - props.percentComplete * 2.55}, ${props.percentComplete * 2.55}, 0, 1)`,
                clipPath: `inset(0 ${100 - props.percentComplete}% 0 0)`
            }}>

            </div>
        </div>
    </div>
}