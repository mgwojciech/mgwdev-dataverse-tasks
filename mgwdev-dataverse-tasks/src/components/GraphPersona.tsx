import { IUser } from "mgwdev-m365-helpers";
import * as React from "react";
import { useGraph } from "../context/GraphContext";
import { PersonaService } from "../services/PersonaService";
import { IPersonaProps, Persona, Spinner } from "@fluentui/react";

export interface IGraphPersonaProps extends IPersonaProps {
  id?: string;
  user?: IUser;
  showPresence?: boolean;
  showSecondaryText?: boolean;
}

export function GraphPersona(props: IGraphPersonaProps) {
  const { id } = props;
  const { graphClient } = useGraph();

  const personaService = React.useRef(new PersonaService(graphClient, props.showPresence))
  const [loading, setLoading] = React.useState(!props.user);
  const [user, setUser] = React.useState<IUser>(props.user);
  const getUserInfo = async () => {
    const userResult = await personaService.current.getUser(id!);
    setUser({
      ...props.user,
      ...userResult,
    });
    setLoading(false);
  };
  React.useEffect(() => {
    if(!props.user)
      getUserInfo();
    else{
      setLoading(false);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [id]);

  React.useEffect(()=>{
    if(props.user){
      setUser(props.user);
      setLoading(false);
    }
  }, [props.user])

  if (loading) {
    return <Spinner />
  }

  const primaryText = user?.displayName || props.text || props.title || props.id;
  return (
    <Persona
      {...props}
      secondaryText={(props.showSecondaryText || props.secondaryText) ? (props.secondaryText || user?.jobTitle) : undefined}
      imageUrl={user?.photo}
    />
  );
}

