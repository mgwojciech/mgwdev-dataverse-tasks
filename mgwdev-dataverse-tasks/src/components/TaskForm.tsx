import * as React from "react";
import { IDataverseTask } from "../model/IDataverseTask";
import { DatePicker, DocumentCard, DocumentCardActivity, DocumentCardStatus, DocumentCardTitle, IconButton, makeStyles, PersonaSize, Slider, TextField } from "@fluentui/react";
import { CompletionIcon } from "./CompletionIcon";
import { TimeZoneService } from "../services/TimeZoneService";
import { GraphPersona } from "./GraphPersona";

export interface ITaskFormProps {
    task?: IDataverseTask;
    onSave?: (task: IDataverseTask) => void;
    onComplete?: (task: IDataverseTask) => void;
    isEditMode?: boolean;
}

const useStyles = makeStyles({
    taskWrapper: {
        display: "flex",
        flexDirection: "column",
        gap: "0.2rem",
        width: "300px",
        padding: "0.5rem",
        border: "1px solid #eaeaea",
        transitionProperty: "all",
        transitionDuration: "0.3s",
        transitionTimingFunction: "ease-in-out",
        "&:hover": {
            boxShadow: "0 0 5px 0 #eaeaea",
            transform: "scale(1.02)"
        }
    },
    taskHeader: {
        display: "flex",
        flexDirection: "row",
        justifyContent: "space-between",
        alignContent: "center",
        alignItems: "center"
    },
    taskDescription: {

    },
    editButton: {

    }
});

export function TaskForm(props: ITaskFormProps) {
    const [task, setTask] = React.useState<IDataverseTask>(props.task || {
        subject: "",
        description: "",
        scheduledstart: new Date().toISOString(),
        scheduledend: new Date().toISOString(),
        percentcomplete: 0
    });
    const [editMode, setEditMode] = React.useState<boolean>(props.isEditMode || false);
    const styles = useStyles();

    React.useEffect(() => {
        setTask(props.task || {
            subject: "",
            description: "",
            scheduledstart: new Date().toISOString(),
            scheduledend: new Date().toISOString(),
            percentcomplete: 0
        });
    }, [props.task])

    React.useEffect(()=>{
        setEditMode(props.isEditMode || false);
    }, [props.isEditMode]);

    return <div className={styles.taskWrapper}>{editMode ? <TaskEditForm {...props} onSave={(task) => {
        setEditMode(false);
        setTask(task);
        props.onSave && props.onSave(task);
    }} /> : <>
        <div className={styles.taskHeader}>{task?.subject}
            <IconButton iconProps={{
                iconName: "Edit",
                onClick: () => {
                    setEditMode(true);
                }
            }} title="Edit" ariaLabel="Edit" /></div>
        <div className={styles.taskDescription}>{task?.description || ""}</div>
        <div>{task && task.scheduledstart && TimeZoneService.getRelativeTime(new Date(task?.scheduledstart), "en-US")}</div>
        <div>{task && task.scheduledend && TimeZoneService.getRelativeTime(new Date(task?.scheduledend), "en-US")}</div>
        <div>{task && <CompletionIcon percentComplete={task.percentcomplete || 0} />}</div>
        <div>{task &&  <IconButton iconProps={{
                iconName: "Accept",
                onClick: () => {
                    setTask({
                        ...task,
                        percentcomplete: 100
                    })
                    props.onComplete && props.onComplete(task);
                }
            }} title="Complete" ariaLabel="Complete" />}</div>
        {/* <div><GraphPersona size={PersonaSize.size32} /></div> */}
    </>}
    </div >
}

export function TaskEditForm(props: ITaskFormProps & { onCancel?: () => void }) {
    const [task, setTest] = React.useState<IDataverseTask>(props.task || {
        subject: "",
        description: "",
        scheduledstart: new Date().toISOString(),
        scheduledend: new Date().toISOString(),
        percentcomplete: 0
    });
    React.useEffect(() => {
        setTest(props.task || {
            subject: "",
            description: "",
            scheduledstart: new Date().toISOString(),
            scheduledend: new Date().toISOString(),
            percentcomplete: 0
        });
    }, [props.task])
    return <div>
        <TextField label="Subject" value={task.subject} onChange={(e, v) => {
            setTest({
                ...task,
                subject: v || ""
            });
        }} />
        <TextField label="Description" value={task.description} onChange={(e, v) => {
            setTest({
                ...task,
                description: v || ""
            });
        }} />
        <Slider label="Percent Complete" min={0} max={100} value={task.percentcomplete} onChange={(v) => {
            setTest({
                ...task,
                percentcomplete: v
            });
        }} />
        <DatePicker label="Start Date" value={new Date(task.scheduledstart)} onSelectDate={(d) => {
            setTest({
                ...task,
                scheduledstart: d?.toISOString() || ""
            });
        }} />
        <DatePicker label="End Date" value={new Date(task.scheduledend)} onSelectDate={(d) => {
            setTest({
                ...task,
                scheduledend: d?.toISOString() || ""
            });
        }} />
        <IconButton iconProps={{
            iconName: "Save",
            onClick: () => {
                props.onSave && props.onSave(task);
            }
        }} title="Save" ariaLabel="Save" />
        <IconButton iconProps={{
            iconName: "Cancel",
            onClick: () => {
                props.onSave && props.onSave(task);
            }
        }} title="Cancel" ariaLabel="Cancel" />
    </div>
}