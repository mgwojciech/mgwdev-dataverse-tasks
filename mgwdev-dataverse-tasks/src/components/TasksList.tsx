import * as React from "react";
import { IDataverseTask } from "../model/IDataverseTask";
import { TasksService } from "../services/TasksService";
import { useDataverse } from "../context/DataverseContext";
import { DataverseUserService } from "../services/DataverseUserService";
import { ActionButton, IconButton, IIconProps, List, makeStyles, Spinner, Text } from "@fluentui/react";
import { TaskForm } from "./TaskForm";

export interface ITasksListProps {

}

const addFriendIcon: IIconProps = { iconName: 'Add' };

const useStyles = makeStyles({
    tasksList: {
        display: "flex",
        flexDirection: "row",
        gap: "1rem",
        flexWrap: "wrap"
    }
});
export const TasksList: React.FC<ITasksListProps> = (props) => {
    const [tasks, setTasks] = React.useState<IDataverseTask[]>([]);
    const [loading, setLoading] = React.useState<boolean>(true);
    const [newTask, setNewTask] = React.useState<IDataverseTask | undefined>(undefined);
    const { dataverseResource, dataverseClient } = useDataverse();
    const service = new TasksService(dataverseClient, dataverseResource!, new DataverseUserService(dataverseClient, dataverseResource!));
    const styles = useStyles();
    React.useEffect(() => {
        service.getMyTasks().then((data) => {
            setTasks(data);
            setLoading(false);
        }).catch((error) => {
            console.error(error);
            setLoading(false);
        });
    }, []);

    return (
        <div>
            <Text as="h2" variant="xLarge">Tasks List</Text>
            <div>
                {loading && <Spinner />}
                {!loading && tasks.length === 0 && <div>No tasks found</div>}
                {!loading && tasks.length > 0 && <div className={styles.tasksList}>{tasks.map((task, index) => <TaskForm
                    key={index}
                    task={task}
                    onComplete={(task) => { service.completeTask(task) }}
                    onSave={(task) => { service.updateTask(task) }}
                />)}
                {newTask && <TaskForm isEditMode={true} task={newTask} onSave={(task) => {
                    service.addTask(task).then(() => {
                        setNewTask(undefined);
                        service.getMyTasks().then((data) => {
                            setTasks(data);
                        });
                    }).catch((error) => {
                        console.error(error);
                    });
                }} />}
                </div>}
            </div>
            <ActionButton iconProps={addFriendIcon} allowDisabledFocus onClick={()=>{
                setNewTask({
                    subject: "",
                    description: "",
                    scheduledstart: new Date().toISOString(),
                    scheduledend: new Date().toISOString(),
                    percentcomplete: 0
                });
            }}>
                Create Task
            </ActionButton>
        </div>
    );
}