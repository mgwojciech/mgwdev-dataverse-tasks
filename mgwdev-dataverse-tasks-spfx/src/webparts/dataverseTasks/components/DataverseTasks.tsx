import * as React from 'react';
import styles from './DataverseTasks.module.scss';
import type { IDataverseTasksProps } from './IDataverseTasksProps';
import { TasksList } from "mgwdev-dataverse-tasks/lib/components/TasksList";

export default class DataverseTasks extends React.Component<IDataverseTasksProps, {}> {
  public render(): React.ReactElement<IDataverseTasksProps> {
 

    return (
      <section className={styles.dataverseTasks}>
        <TasksList />
      </section>
    );
  }
}
