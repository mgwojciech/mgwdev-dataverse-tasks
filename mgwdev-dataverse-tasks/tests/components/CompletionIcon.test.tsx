import { vi, expect } from "vitest";
import * as React from "react";
import { CompletionIcon, ICompletionIconProps } from "../../src/components/CompletionIcon";
import { render } from "@testing-library/react";

describe("CompletionIcon", () => {
    test("should render completion icon (0%)", () => {
        const props: ICompletionIconProps = {
            percentComplete: 0
        };
        const { container } = render(<CompletionIcon {...props} />);
        expect(container).toMatchSnapshot();
    });
    test("should render completion icon (50%)", () => {
        const props: ICompletionIconProps = {
            percentComplete: 50
        };
        const { container } = render(<CompletionIcon {...props} />);
        expect(container).toMatchSnapshot();
    });
    test("should render completion icon (100%)", () => {
        const props: ICompletionIconProps = {
            percentComplete: 100
        };
        const { container } = render(<CompletionIcon {...props} />);
        expect(container).toMatchSnapshot();
    });
});