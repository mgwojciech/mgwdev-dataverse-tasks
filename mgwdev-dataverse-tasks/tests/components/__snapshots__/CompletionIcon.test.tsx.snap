// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CompletionIcon > should render completion icon (0%) 1`] = `
<div>
  <div
    class="root-109"
  >
    <div
      class="wrapper-110"
    >
      <div
        class="circle-111"
        style="border-color: rgba(255, 0, 0, 1); clip-path: inset(0 100% 0 0);"
      />
    </div>
  </div>
</div>
`;

exports[`CompletionIcon > should render completion icon (50%) 1`] = `
<div>
  <div
    class="root-109"
  >
    <div
      class="wrapper-110"
    >
      <div
        class="circle-111"
        style="clip-path: inset(0 50% 0 0);"
      />
    </div>
  </div>
</div>
`;

exports[`CompletionIcon > should render completion icon (100%) 1`] = `
<div>
  <div
    class="root-109"
  >
    <div
      class="wrapper-110"
    >
      <div
        class="circle-111"
        style="clip-path: inset(0 0% 0 0);"
      />
    </div>
  </div>
</div>
`;
